File: python/src/Core/Logger.py
import json
import os
import secrets
from datetime import datetime, timezone, timedelta

class Logger:
    log_file = os.path.join(os.path.dirname(__file__), "../../storage/logs/app.log")

    @staticmethod
    def info(message, meta_data=None, data=None):
        Logger.write("INFO", message, meta_data or {}, data or {})

    @staticmethod
    def error(message, meta_data=None, data=None):
        Logger.write("ERROR", message, meta_data or {}, data or {})

    @staticmethod
    def write(level, message, meta_data, data):
        log = {
            "timestamp": Logger.timestamp(),
            "message": message,
            "level": level,
            "trace_id": Logger.trace_id(),
            "service": "users-api",
            "meta_data": meta_data,
            "data": data,
        }

        with open(Logger.log_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(log, ensure_ascii=False) + "\n")

    @staticmethod
    def trace_id():
        return f"{secrets.token_hex(8)},{secrets.token_hex(8)}"

    @staticmethod
    def timestamp():
        now = datetime.now().astimezone()
        return now.strftime("%Y%m%dT%H:%M:%S%z")[:-2] + ":" + now.strftime("%Y%m%dT%H:%M:%S%z")[-2:]

File: javascript/src/Core/Logger.js
const fs = require('fs');
const path = require('path');

const logFile = path.join(__dirname, '../../storage/logs/app.log');

class Logger {

  static info(message, metaData = {}, data = {}) {
    this.write("INFO", message, metaData, data);
  }

  static error(message, metaData = {}, data = {}) {
    this.write("ERROR", message, metaData, data);
  }

  static write(level, message, metaData = {}, data = {}) {
    const log = {
      timestamp: this.formatTimestamp(),
      message,
      level,
      trace_id: this.generateTraceId(),
      service: "users-api",
      meta_data: metaData,
      data: data,
    };

    fs.appendFileSync(logFile, JSON.stringify(log) + "\n");
  }

  static generateTraceId() {
    return (
      this.randomHex(16) + ',' + this.randomHex(16)
    );
  }

  static randomHex(length) {
    return [...Array(length)]
      .map(() => Math.floor(Math.random() * 16).toString(16))
      .join('');
  }

  static formatTimestamp() {
    const now = new Date();

    const tzOffset = -now.getTimezoneOffset();
    const sign = tzOffset >= 0 ? '+' : '-';
    const hours = String(Math.floor(Math.abs(tzOffset) / 60)).padStart(2, '0');
    const minutes = String(Math.abs(tzOffset) % 60).padStart(2, '0');

    return now.toISOString().replace(/[-:]/g, '').split('.')[0] + sign + hours + ':' + minutes;
  }
}

module.exports = Logger;